mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ExecutorStateManagerImpl.java

```diff
@@ -63,55 +69,63 @@ public class ExecutorStateManagerImpl implements ExecutorStateManager {
         .build();
 
     @Override
-    public void putIfAbsent(TaskExecutorID taskExecutorID, TaskExecutorState state) {
+    public boolean onTaskExecutorStateAssigned(TaskExecutorID taskExecutorID, TaskExecutorState state) {
         this.taskExecutorStateMap.putIfAbsent(taskExecutorID, state);
         if (this.archivedState.getIfPresent(taskExecutorID) != null) {
             log.info("Reviving archived executor: {}", taskExecutorID);
             this.archivedState.invalidate(taskExecutorID);
         }
 
-        // add to buckets. however new executors won't have valid registration at this moment and requires marking
-        // again later when registration is ready.
+        return tryMarkAvailable(taskExecutorID, state);
+    }
+
+    /**
+     * Add to buckets. however new executors won't have valid registration at this moment and requires marking
+     * again later when registration is ready.
+     * @param taskExecutorID taskExecutorID
+     * @param state state
+     * @return whether the target executor is marked as available.
+     */
+    public boolean tryMarkAvailable(TaskExecutorID taskExecutorID, TaskExecutorState state) {
         if (state.isAvailable() && state.getRegistration() != null) {
-            log.info("Marking executor {} as available for matching.", taskExecutorID);
+            TaskExecutorHolder teHolder = TaskExecutorHolder.of(taskExecutorID, state.getRegistration());
+            log.debug("Marking executor {} as available for matching.", teHolder);
             double cpuCores = state.getRegistration().getMachineDefinition().getCpuCores();
             if (!this.executorByCores.containsKey(cpuCores)) {
-                this.executorByCores.putIfAbsent(cpuCores, new HashSet<>());
+                this.executorByCores.putIfAbsent(
+                    cpuCores,
+                    new TreeSet<>(TaskExecutorHolder.generationFirstComparator));
             }
 
-            this.executorByCores.get(cpuCores).add(taskExecutorID);
+            log.info("Assign {} to available.", teHolder.getId());
+            return this.executorByCores.get(cpuCores).add(teHolder);
+        }
+        else {
+            log.warn("Ignore unavailable TE: {}", taskExecutorID);
+            return false;
         }
     }
 
     @Override
-    public void markAvailable(TaskExecutorID taskExecutorID) {
+    public boolean tryMarkAvailable(TaskExecutorID taskExecutorID) {
         if (!this.taskExecutorStateMap.containsKey(taskExecutorID)) {
             log.warn("marking invalid executor as available: {}", taskExecutorID);
-            return;
+            return false;
         }
 
         TaskExecutorState taskExecutorState = this.taskExecutorStateMap.get(taskExecutorID);
-        if (taskExecutorState.getRegistration() == null) {
```

Comment:
It looks like with the old logic we would allow TEs with non-available state to be added to `executorByCores`. I'm wondering why.